frontend/document/credocument.lua

```diff
@@ -1370,50 +1372,53 @@ function CreDocument:setupCallCache()
 
     -- reset full cache
     self._callCacheReset = function()
-        self._call_cache = {}
-        self._call_cache_tags_lru = {}
+        -- "Global" cache, a simple key, value store for *this* document
+        self._call_cache_global = {}
+        -- "Tags" cache, an LRU cache of per-tag simple key, value stores. 10 slots.
+        if self._call_cache_tags then
+            self._call_cache_tags:clear()
+        else
+            self._call_cache_tags = lru.new(10)
+        end
+        -- i.e., the only thing that follows any sort of LRU eviction logic is the *list* of tag caches.
+        -- Each individual cache itself is just a simple key, value store (i.e., a hash map).
+        -- Points to said per-tag cache for the current tag.
+        self._call_cache_tag = nil
+        -- Stores the key for said current tag
+        self._call_cache_current_tag = nil
     end
     -- global cache
     self._callCacheGet = function(key)
-        return self._call_cache[key]
+        return self._call_cache_global[key]
     end
     self._callCacheSet = function(key, value)
-        self._call_cache[key] = value
+        self._call_cache_global[key] = value
     end
 
-    -- nb of by-tag sub-caches to keep
-    self._call_cache_keep_tags_nb = 10
     -- current tag (page, pos) sub-cache
     self._callCacheSetCurrentTag = function(tag)
-        if not self._call_cache[tag] then
-            self._call_cache[tag] = {}
+        -- If it already exists, return it and make it the MRU
+        self._call_cache_tag = self._call_cache_tags:get(tag)
+        if not self._call_cache_tag then
+            -- Otherwise, create it and insert it in the list cache, evicting the LRU tag cache if necessary.
+            -- NOTE: We need CacheItem for its NOP onFree eviction callback.
+            self._call_cache_tag = CacheItem:new{}
```

Comment:
~~Why does this needs to be a CacheItem, and just not a simple table?~~
Oh, you wrote why :)

Still, can you explain _for its NOP onFree eviction callback._ ?
(We're only storing values that don't need any freeing, aren't we?)

Reply:
Yeah, but ffi/lru (currently) calls it unconditionally.

I'll see if I can deal with that in a saner manner tomorrow (possibly by template-ing the del function based on an instantiation setting).

Reply:
Fixed it ;).